### ISSUE TITLE
Figuring out our strategy for tabbed content

### ISSUE THREAD

**Issue by pradyunsg** [executablebooks/meta#145]
**Is your feature request related to a problem? Please describe.**

Currently, sphinx-tabs and sphinx-panels both provide tabbed-content functionality. This overlap isn't super useful and creates a very weird dynamic. Right now, we have two implementations of the same fundamental thing living in 2 packages maintained by the same group of people.

- sphinx-tabs: currently outdated, is generally more capable than sphinx-panel's CSS-only implementation of tabs.
- sphinx-panels: A CSS-only package, providing (among other things) a `tabbed` directive that generates tabs adopted from mkdocs-material, looks *great* on the final page. :)

**Describe the solution you'd like**

We move the sphinx-panels implementation of tabs into sphinx-tabs and remove the `tabbed` directive from sphinx-panels. The sphinx-tabs implementation would be modernized, and, heck, we could even have the tabbed directive added to it directly.

Basically, I'm volunteering to rewrite sphinx-tabs to look nicer whilst also adding graceful degradation and solving executablebooks/sphinx-tabs#87 (and a few more) but I don't want to step on anyone's toes and I want to make sure we're all on the same page. :)

**Open Questions**

- What does Chris think about this idea? :P
- Directive design:
  - How important is it to preserve the current structure of sphinx-tabs's directives?
  - How important is it to preserve the current structure of the `tabbed` directive?
  - Should we include both or only one of them in the rewrite? If so, which one?

**Comment by chrisjsewell** [comment]
Well I'm not really that bothered about sphinx-tabs anymore, now that (simple) tabs are in sphinx-panels lol, so go ahead do your worst (although we should pay some attention to back compatibility for current users)
As to whether this should then prompt the removal of `tabbed` from sphinx-panels.
It really just depends on the final implementation. sphinx-panels is now included in jupyter-book and provided a nice "non-obtrusive" addition for general users, and obviously I named it different to the directives in `sphinx-tabs` so that both can still be used without issue.
I would be hesitant to include sphinx-tabs at least in its current form.

**Comment by choldgraf** [comment]
yeah I think it's worth emphasizing that last point - `sphinx-tabs` isn't bundled with Jupyter Book, and so I think that reduces the confusion a bit. I think of it as more of a "power user" tab tool that we can also expect users to be able to manually configure if they wish.

I do think that at some point if panels continues growing in complexity, we'll want to modularize it a bit, and perhaps this could be a part of that, but I don't think it's a big fire right now

@pradyunsg what are the main features that `sphinx-tabs` has that this implementation does not?

**Comment by pradyunsg** [comment]
@choldgraf Synchronized tabs, and... that's it? Yea. :)

**Comment by choldgraf** [comment]
I wonder if @chrisjsewell would be willing to just baaaaarely relax his "no JS" rule in this one use-case of synchronized tabs? :-)

**Comment by chrisjsewell** [comment]
> I wonder if @chrisjsewell would be willing to just baaaaarely relax his "no JS" rule in this one use-case of synchronized tabs? :-)

I wouldn't hold your breath. First you would have to actually provide a PR for this solution, and also show that you have tried to implement it in a non-JS way first.
Even then I'm not guaranteeing that I would accept the PR. I have already explained the benefits of CSS only approaches over JS, and there are even other benefits that I hadn't considered (see https://github.com/executablebooks/sphinx-tabs/issues/92). If you are going to break all that for a feature that no one on jupyter-book has actually asked for yet, then I would politely decline.

**Comment by chrisjsewell** [comment]
never say never though 😆

**Comment by choldgraf** [comment]
I am on board with @chrisjsewell 's general principle of no JS unless absolutely necesssary. I'm also not a fan of maintaining two libraries with 90% overlap in functionality (tabs in panels vs. sphinx-tabs), so really my main point is that I'd like to find a way for us to deprecate one of those two pieces :-)  (probably sphinx-tabs?)

**Comment by pradyunsg** [comment]
From my perspective... the tabs implementation in sphinx-panels is basically independent of the rest of sphinx-panels. It doesn't depend on bootstrap, or on any other elements in sphinx-panels.

All the deficiencies of sphinx-tabs that @chrisjsewell is flagging are... well, trivial to fix IMO. That isn't to say they aren't deficiencies or unimportant to address, but rather that I view them as a consequence of the status-quo implementation there not being great/not considering those needs and not a case of CSS-only vs JS+CSS.

A rewrite based off of what sphinx-panels is doing + a touch of JS would be much better on all those fronts IMO. 

> See if we can implement sync-ed tabs with CSS and HTML only

I tried this and gave up - there's an issue on sphinx-panels I filed about synchronized tabs, that has the story of me trying and giving up.

> Only merge if it is maybe < 10 lines of JS or something.

It'll solve the problem, but... I don't see good reasons to do it this way. What are the pros of doing things this way?

One of the cons is that we'd basically have a component (tabs) that works without any of the others, works without bootstrap, that is provided as part of a library that defaults to you're gonna use bootstrap.

And we're gonna end up deprecating an existing package with an existing userbase for exactly the same functionality that... well... has a lot of scope for improvement but doesn't look like there's any deal breakers tbh. :)

> no one on jupyter-book has actually asked for yet

Welllll, I can't speak for jupyter-book users, but pip's documentation could realllly use some good looking synchronized tabs per-OS. I'm pretty sure that lots of Python projects might want this.

Right now, the choices are "old meh-looking heavy tabs based on semantic-UI" vs "clean looking tabs that don't depend on anything w/ added weight from everything else sphinx-panels provides". For a component that "just works" independently... Idk if I like any of these options.

---

> bring "sphinx-panels tabs into sphinx-tabs"

In spirit of moving things forward, is anyone actively opposed to this idea? :)

This would mean rewriting sphinx-tabs, which shouldn't take more than 1 evening IMO, and deprecating the tabbed directive in sphinx-panels in favor of sphinx-tabs. FWIW, I'm pretty open to also adding that directive to sphinx-tabs, but we'll cross that bridge once we get there. :)

**Comment by akhmerov** [comment]
Perhaps this is a decent gauge in judging demand for grouped tabs: https://github.com/search?l=reStructuredText&q=group-tab&type=code

**Comment by choldgraf** [comment]
> In spirit of moving things forward, is anyone actively opposed to this idea? :)

I'm not opposed to it. I didn't realize that `sphinx-tabs` had such a big userbase already. Given that fact, I'd be +1 on moving this into `sphinx-tabs`.

**Comment by chrisjsewell** [comment]
We can discuss it again when there is a viable alternative on the table.
Until that day, the `tabbed` directive will remain in sphinx-panels.

**Comment by choldgraf** [comment]
I think @pradyunsg proposed this alternative:

> This would mean rewriting sphinx-tabs, which shouldn't take more than 1 evening IMO, and deprecating the tabbed directive in sphinx-panels in favor of sphinx-tabs.

AKA, re-write the current `sphinx-tabs` directives to be more lightweight and use the CSS approach that the `sphinx-panels` implementation uses.

And in addition, if the `tabbed` directive pattern is something we want to keep:

> FWIW, I'm pretty open to also adding that directive to sphinx-tabs, but we'll cross that bridge once we get there.

**Comment by chrisjsewell** [comment]
No by "on the table", I mean someone has actually written code.

Note we also have to consider that, given sphinx-tabs does have a biggish user-base, if those users are going to be happy if sphinx-tabs suddenly completely changes how it looks and perhaps works.

TBH it seems that @pradyunsg has a very specific design in mind, so I don't why you don't just write a new sphinx extension.
There's no rule against having multiple extensions with the same functionality, that users can choose between, but likely once this is in existence then we can decide if/where it should get moved / what it should replace.

**Comment by choldgraf** [comment]
So it sounds like for the purposes of this issue our strategy for tabbed content is:

- There are two sphinx extensions that handle tabbed content, `sphinx-panels` and `sphinx-tabs`. While they are both under `executablebooks/`, there is no direct plan to integrate them
- If somebody wants to re-write `sphinx-tabs` to be more streamlined, and replicates the functionality of `sphinx-panels`, then `sphinx-panels` *may* consider deprecating in favor of `sphinx-tabs` but no guarantees there
- Jupyter Book installs and documents `sphinx-panels` functionality on its own, and will recommend users use `sphinx-tabs` if they want more advanced tabs functionality (like synced tabs)

**Comment by pradyunsg** [comment]
Thanks a lot for aggregating that @choldgraf! ^>^

> TBH it seems that @pradyunsg has a very specific design in mind, so I don't why you don't just write a new sphinx extension.

Welll... I didn't feel like I did, but the more I thought about it, the more this felt true! So... I made it a thing: https://github.com/pradyunsg/sphinx-inline-tabs! Docs+Demo at https://sphinx-inline-tabs.readthedocs.io/en/latest/.

With that, my wish for a small package that is only-tabs-and-does-it-well-enough is met. So... I'm no longer planning to rewrite `sphinx-tabs`. I did start that route, but two evenings in, I realized that I was dealing with its current design's backwards-compatibility implications and me-no-wanna-deal-with-that.

If anything, I guess I'd request that `sphinx-inline-tabs` be added to the recommended alternatives in Jupyter Book and sphinx-panels (in addition to `sphinx-tabs`) since it's basically super similar to sphinx-panels with a tiny bit of JS to make stuff stay synchronized.

Other than that... hmm... nothing else really! This discussion was super useful folks! Thanks a lot! ^>^

**Comment by choldgraf** [comment]
awesome! very nice!

so just to summarize we now have:

- `sphinx-tabs` which is a heavy-weight tabs implementation that also has a large user-base
- `sphinx-panels` tabs which is a super lightweight implementation of tabs with a "sequential-directives" approach and no JS
- `sphinx-inline-tabs` is a somewhat lightweight implementation of tabs that also has a "sequential-directives" approach and uses some JS

**Comment by pradyunsg** [comment]
I'd phrase it as:

- sphinx-tabs' has nesting and uses `tabs`, `tab`, `group-tab` and `code-tab`.
- sphinx-panel's `tabbed` works approximately like `tab` from sphinx-tabs, and is sequential. 
- sphinx-inline-tabs' `tab` works approximately like `group-tab` from sphinx-tabs, and is sequential as well.

**Comment by akhmerov** [comment]
Inline tabs is absolutely wonderful: both the appearance and the markup design!

**Comment by choldgraf** [comment]
gotcha - thanks for the explanation! I quite like all three for various reasons :-) It's unfortunate that now there will be three different "tab" syntaxes out there, but I suppose it wouldn't be that hard for people to move between the projects by doing a find/replace for the directive name

**Comment by pradyunsg** [comment]
> both the appearance and the markup design!

Credit to @chrisjsewell -- I basically copied and modified what they did in sphinx-panels. lol

---

> It's unfortunate that now there will be three different "tab" syntaxes out there

YEA. I would've have loved to use `tabbed` for sphinx-inline-tabs (since it's basically 80% compatible w/ sphinx-panel's implementation) but that'd mean it can't be used with sphinx-panels without some hackery, sooooooo, it's maybe better this way. :)

> I suppose it wouldn't be that hard for people to move between the projects by doing a find/replace for the directive name

Yup. The switch in Furo's docs (which use tabs heavily in the [Reference](https://pradyunsg.me/furo/reference/) section) was literally find-replace `tabbed` -> `tab`.

Between sphinx-tabs and the rest... it'd be more tricky. :)

**Comment by choldgraf** [comment]
yeah, I am comfortable recommending to people "you should use one of these two packages instead of `sphinx-tabs`...and I agree with everyone's intuition about not deprecating syntax that is already super heavily-used.

**Comment by chrisjsewell** [comment]
FYI @pradyunsg i just opened the page on my iPhone, and the syncing didn't work. (it does for sphinx-tabs)

**Comment by choldgraf** [comment]
weird - it doesn't sync for me either on Chrome (though it did before when I tried).

JS has this error:

![image](https://user-images.githubusercontent.com/1839645/96054731-2a43ab00-0e37-11eb-9ca2-971e9f3750bd.png)

**Comment by pradyunsg** [comment]
Yea, I borked it earlier today -- see https://github.com/pradyunsg/sphinx-inline-tabs/issues/1. I'll come back to it later.

**Comment by choldgraf** [comment]
haha `#opensource`

**Comment by foster999** [comment]
Great work on `sphinx-inline-tabs` - looks like a good combination of features from the others!

Are there are any comprehensive docs for using `docutils`? Ironically most of their methods don't seem to have descriptions, so I've hashed executablebooks/sphinx-tabs#94 together through copying snippets from elsewhere + trial and error... There's certainly some improvements I can make from looking at the `sphinx-panels` and  `sphinx-inline-tabs`  implementations.

**Comment by chrisjsewell** [comment]
Well there's https://docutils.sourceforge.io/, which is pretty rubbish.
I've already had a bit of an argument with them over moving from sourceforge to GitHub (and adding RTD docs), but they are just not interested: https://sourceforge.net/p/docutils/feature-requests/58/

See also: https://github.com/sphinx-doc/sphinx/issues/8039

**Comment by foster999** [comment]
I thought that might be the case 😢

Hope they take you up on the migration sometime!

**Comment by pradyunsg** [comment]
Gah, that docutils discussion makes me sad. I'mma say we're done with this conversation tho -- with https://github.com/executablebooks/meta/issues/145#issuecomment-708671652 serving as the summary of status quo.

Happy to let someone else close this, assuming they agree. ;)

**Comment by choldgraf** [comment]
That makes sense for me. What I'd like to do is, as a start, remove the `sphinx-tabs` example here: https://jupyterbook.org/advanced/sphinx.html#an-example-sphinx-tabs

and replace it with the same idea but with `sphinx-inline-tabs`. @pradyunsg would you be interested in implementing this?

**Comment by chrisjsewell** [comment]
Ps
I note also that your CSS and JS doesn't have any hashing: booo 😜 
Plus why not just release with a GH action: here's one using flit: https://github.com/executablebooks/mdformat-tables/blob/c2855ea749fa4b8974efcd97c307cf96becc40b0/.github/workflows/tests.yml#L77

**Comment by pradyunsg** [comment]
> Yea, I borked it earlier today

https://github.com/pradyunsg/sphinx-inline-tabs/commit/d17680c111c472f9929ce6791266e483d24647a2 -- it's fixed and me feels stupid.

>  I note also that your CSS and JS doesn't have any hashing: booo 😜

I doubt I'd even be updating the package much, since "keep it small and simple" is kind of the point, and beyond accessibility, I don't see how things would really change enough to justify this additional bit of code.

**Comment by pradyunsg** [comment]
> @pradyunsg would you be interested in implementing this?

I missed this -- yes, I would be! OTOH, I'm moving countries and have a bit of a backlog on pip's 20.3 release that I have to get to, so I'd be a happy kiddo if someone else gets to it before I do. :)
